### PR TITLE
docs: provide a guide for how to use IBM Secure Execution

### PR DESCRIPTION
**Pull request by BbolroC**
This PR is to add a document for how to run kata containers under IBM Secure Execution environment.

Fixes: #7025

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

### CODE REVIEW

**Comment by stevenhorsman** [comment]
Hey Choi this is failing with the error:
`ERROR: Document docs/how-to/how-to-run-kata-containers-with-SE-VMs.md is not referenced`, so I would suggest adding a link to the doc in `docs/how-to/README.md`

**Comment by BbolroC** [comment]
> Just a thought - I wonder if the CoCo operator section would be better off in the coco documentation. Maybe as a new entry in https://github.com/confidential-containers/documentation/tree/main/guides like sev has? It might be more appropriate to document/link to the [kata-deploy](https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/README.md) in this doc?

Yeah, I will let you know after discussing what you mentioned with @hbrueckner . Thanks! 😉

**Comment by BbolroC** [comment]
@hbrueckner I have relocated the section for CC operator in https://github.com/confidential-containers/documentation/pull/133 and made this document refer to the kata-deploy.

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+To leverage the IBM Secure Execution (SE) capability, the host machine on which you intend
+to run workloads must be an IBM z15 (or a newer model) or an IBM LinuxONE III (or a newer
+model). In addition to the hardware requirement, you need to verify the CPU facility and
+kernel configuration, as outlined below:
+
+```
+$ # To check the protected virtualization support from kernel
+$ grep -q 'prot_virt=1' /etc/zipl.conf && echo "found" || echo "not found"
+found
+$ # To check if an ultravior reserves a memory for the current boot
```
Comment (zvonkok) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:21]:
```suggestion
$ # To check if an ultravisor reserves a memory for the current boot
```

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+$ # Change a directory to the project root
+$ cd $GOPATH/src/github.com/kata-containers/kata-containers
+$ host_key_document=$HOME/host-key-document/HKD-0000-0000000.crt
+$ kernel_image=artifacts/opt/kata/share/kata-containers/vmlinuz-5.19.2-107
+$ initrd_image=artifacts/opt/kata/share/kata-containers/kata-ubuntu-20.04.initrd
+$ echo "panic=1 scsi_mod.scan=none swiotlb=262144 agent.log=debug" > parmfile
+$ sudo genprotimg --host-key-document=${host_key_document} \
+--output=kata-containers-se.img --image=${kernel_image} --ramdisk=${initrd_image} \
+--parmfile=parmfile --no-verify
+WARNING: host-key document verification is disabled. Your workload is not secured.
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:127]: It might be worth to also add some steps with verification.

@BbolroC Could you talk with @NicoBoehrIBM (talked with him already offline)?

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+$ file kata-containers-se.img
+kata-containers-se.img: data
+$ sudo cp kata-containers-se.img /opt/kata/share/kata-containers/
+```
+
+The aforementioned steps, including the dependencies for the kernel and initrd,
+can be easily accomplished by executing the following make target:
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:134]: style: `running` instead of `executing`

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+```
+
+Finally, an operational kata container with IBM SE is now running. 
+
+## Using Kata-Deploy
+
+It is reasonable to anticipate that the aforementioned manual steps can be readily
+accomplished by the
+[kata-deploy](https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/README.md).
+The following commands explains how to build an archive named `kata-static.tar.xz`:
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:235]: typo: `explain` (plural)

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+$ sudo -E PATH=$PATH make shim-v2-tarball
+$ mkdir kata-artifacts
+$ build_dir=$(readlink -f build)
+$ sudo cp -r $build_dir/*.tar.xz kata-artifacts
+$ sudo chown -R $(id -u):$(id -g) kata-artifacts
+$ ./tools/packaging/kata-deploy/local-build/kata-deploy-merge-builds.sh kata-artifacts
+```
+
+If a rootfs-image is required for testing purposes without SE functionality,
+it is necessary to execute the following command
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:256]: style: `run` instead of `execute`

**Review by hbrueckner — COMMENTED** [review]
Hi @BbolroC ,

thanks for this excellent documentation. I have only a few edits and remarks to follow-up. Thanks a lot!

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+1. Host capable of Secure Execution
+
+To take advantage of the IBM Secure Execution capability, the host machine on which
+you intend to run workloads must be an IBM z15 (or a newer model) or an IBM LinuxONE III
+(or a newer model). In addition to the hardware requirement, you need to verify the CPU
+facility and kernel configuration, as outlined below:
+
+```
+$ # To check the protected virtualization support from kernel
```
Comment (NicoBoehrIBM) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:18]: I think this comment may be misleading. If this returns "found" it simply means the parameter is present. You can, however, still have a kernel which doesn't support SE or a kernel that supports SE, but doesn't have the parameter.
I see two possibilities to make this clearer:
1. If you want to check whether your kernel in theory supports SE, you can check whether `/sys/firmware/uv/prot_virt_host` is present. If it contains `1`, the `prot_virt` parameter was set and the ultravisor was initialized.
1. If you just want to see whether the parameter is there, maybe rephrase the comment to say "Check if the kernel is configured as a SE host"

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+you can construct a secure image using the following procedure:
+
+```
+$ # Change a directory to the project root
+$ cd $GOPATH/src/github.com/kata-containers/kata-containers
+$ host_key_document=$HOME/host-key-document/HKD-0000-0000000.crt
+$ kernel_image=artifacts/opt/kata/share/kata-containers/vmlinuz-5.19.2-107
+$ initrd_image=artifacts/opt/kata/share/kata-containers/kata-ubuntu-20.04.initrd
+$ echo "panic=1 scsi_mod.scan=none swiotlb=262144 agent.log=debug" > parmfile
+$ sudo genprotimg --host-key-document=${host_key_document} \
```
Comment (NicoBoehrIBM) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:122]: Do we really need `sudo` here? Also below there are a few commands which look like they would work fine without `sudo`. For `genprotimg`, I would strongly recommend running without sudo, for the rest of the commands I'm not sure.

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+### Build a Secure Image
+
+
+Assuming you have placed the following files at `$HOME/host-key-document`:
+
+- Host key document as `HKD-0000-0000000.crt`
+- DigiCert CA certificate as `DigiCertCA.crt`
+- IBM Z signing-key certificate as `ibm-z-host-key-signing.crt`
+
+you can construct a secure image using the following procedure:
```
Comment (NicoBoehrIBM) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:128]: Maybe we emphasize that this is not secure:

you can construct a image for testing purposes using the following procedure (not secure, see below for the production ready commands):

**Review by NicoBoehrIBM — COMMENTED** [review]
Besides one small nit, LGTM

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md (hunk 8 above)
Comment (BbolroC) [reply docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:128]: Okay, I will update the doc. Thanks for the comment. 😉

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+$ git clone -b $tool_version https://github.com/ibm-s390-linux/s390-tools.git && cd s390-tools
+$ pushd genprotimg && make && sudo make install && popd
+```
+
+4. Host Key Document
+
+A host key document is a public key employed for encrypting a secure image, which is
+subsequently decrypted using a corresponding private key during the VM bootstrap process.
+You can obtain the host key document either through IBM's designated
+[resource link](http://www.ibm.com/servers/resourcelink) or by requesting it from the
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:107]: minor: use "Resource Link" (as shown on the website)

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+cloud provider responsible for the IBM Z and LinuxONE instances where your workloads are intended to run.
+
+To ensure security, it is essential to verify the authenticity and integrity of the host key document
+belonging to an authentic IBM machine. To achieve this, please additionally obtain the following
+certificates from the resource link:
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:112]: ... `from Resource Link`:

**Review by hbrueckner — COMMENTED** [review]
just two minor edits

**Review by wainersm — APPROVED** [review]
Hi @BbolroC, this doc helped me to understand some decisions you took on https://github.com/kata-containers/kata-containers/pull/6755 so helped me on the review too!

**Comment by BbolroC** [comment]
For reviewers, you can find the referred documentation in the section `Using Kata-Deploy with Confidential Containers Operator` at https://github.com/confidential-containers/operator/pull/329

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+Please execute the following command to run a container:
+
+```
+$ cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx-kata
+spec:
+  runtimeClassName: kata-qemu
```
Comment (stevenhorsman) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:278]: Hi Choi, this is probably a stupid question, but where does `kata-qemu-se` runtimeClass come into use over `kata-qemu`. Is it just when using kata-deploy rather than the manual approach as the other runtime class hasn't been created?

Comment (BbolroC) [reply docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:278]: The document starts with an assumption that kata container is installed. It would be good to have how to generate a new runtime class like `kata-qemu-se` to the doc, but I just wanted to make it focus on explaining how to run a kata container on IBM Z SE based on the basic setup. 

And yes, you are right, a runtime class `kata-qemu-se` is introduced in the doc for confidential containers. 😉

**Comment by BbolroC** [comment]
https://github.com/kata-containers/kata-containers/actions/runs/7585438489/job/20661329265?pr=7146 will be resolved when https://github.com/confidential-containers/operator/pull/329 is merged.

**Review by stevenhorsman — APPROVED** [review]
I don't have a spare z16 LPAR to test on, but I've read through the documentation and it looks good to me. Thanks!

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+[    0.063630] prot_virt.f9efb6: Reserving 98MB as ultravisor base storage
+$ # To check a facility bit for Secure Execution
+$ cat /proc/cpuinfo | grep 158
+facilities      :  ... numbers ... 158 ... numbers ...
+```
+
+If any of the results are not identifiable, please reach out to the responsible cloud
+provider to enable the Secure Execution capability. Alternatively, if you possess
+administrative privileges and the facility bit is set, you can enable the Secure Execution
+capability by adding `prot_virt=1` to the kernel parameters and performing a system reboot with:
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:41]: minor: ...performing a system reboot_, for example,_ with:....

The example is using the `zipl.conf` directly... there are meanwhile zipl support for BLS which need to be differently configured. So just make this as an example, as distros might vary.

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+Typically, you can use
+[kata-deploy](https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/README.md)
+to install Kata Containers on a Kubernetes cluster. However, when leveraging IBM Secure Execution,
+you need to employ the confidential container's
+[operator](https://github.com/confidential-containers/operator).
+During this process, a `kata-deploy` container image serves as a payload image in a custom
+resource `ccruntime` for confidential containers, enabling the operator to install Kata
+binary artifacts such as kernel, shim-v2, and more.
+
+In this section, we will specifically guide you on building a payload image
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:313]: minor: remove "we" and use "you will learn" or "This section explains how" / "This section guides you through....."

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+91462f44bb06: Layer already exists
+2ad49fac591a: Layer already exists
+5c75aa64ef7a: Layer already exists
+test: digest: sha256:25825c7a4352f75403ee59a683eb122d5518e8ed6a244aacd869e41e2cafd385 size: 1369
+```
+
+## Considerations for CI
+
+If you intend to integrate the aforementioned procedure with a CI system, it is
+recommended to configure the following setup for an environment variable.
```
Comment (hbrueckner) [inline docs/how-to/how-to-run-kata-containers-with-SE-VMs.md:381]: minor: remove "it is recommended" and simply use "configure"

**Review by hbrueckner — APPROVED** [review]
Hi Choi,
thanks a lot for pulling this excellent guide together. I have only really minor remarks/edits that should not hold back any approval.

/approve

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md (hunk 12 above)
Comment (BbolroC) [reply docs/how-to/how-to-run-kata-containers-with-SE-VMs.md]: Thanks for the feedback. The document has been updated.

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md (hunk 13 above)
Comment (BbolroC) [reply docs/how-to/how-to-run-kata-containers-with-SE-VMs.md]: Thanks for the feedback. The document has been updated.

File: docs/how-to/how-to-run-kata-containers-with-SE-VMs.md (hunk 14 above)
Comment (BbolroC) [reply docs/how-to/how-to-run-kata-containers-with-SE-VMs.md]: Thanks for the feedback. The document has been updated.
